bosonprotocol/boson-protocol-contracts
# Nested structs Part 1: Associate Condition with Group

## Group / Condition refactor

- Eliminate the storage of nested structs
- Associate Condition with Group rather than including as a struct member

### Changes
* In BosonTypes.sol,
  - in removed condition member from Group

* In ExchangeHandlerFacet.sol,
  - in authorizeCommit method
    - fetch the condition
    - refer to condition not group.condition throughout

* In GroupBase.sol,
  - in createGroupInternal method
    - add _condition argument
    - refer to _condition not _group.condition throughout
    - instead of setting group.condition, call new storeCondition method
    - add _condition to the emitted GroupCreated event
  - added storeCondition method

* In GroupHandlerFacet.sol,
  - in createGroup method,
    - add _condition argument
    - pass _condition to createGroupInternal
  - in removeOffersFromGroup method,
    - fetch the condition
    - add condition to the emitted GroupUpdated event
  - in setGroupCondition method,
    - call storeCondition rather than setting group.condtion
    - add _condition to emitted GroupUpdated event
  - in getGroup method,
    - added condition return parameter
    - fetch the group and if it exists fetch the condition

* In IBosonGroupEvents.sol,
  - added condition to GroupCreated and GroupUpdated events

* In IBosonGroupHandler.sol,
  - in createGroup method
    - added _condition argument
  - in getGroup method,
    - added condition return parameter

* In OrchestrationHandlerFacet.sol,
  - in all methods taking a _condition arg
    - changed _condition arg to use calldata location
  - in createOfferWithCondition method,
    - pass _condition to createGroupInternal

* In ProtocolBase.sol,
  - added internal fetchCondition method

* In ProtocolLib.sol,
  - added ProtocolEntities.conditions mapping

### Tests
* In Group.js,
  - remove condition from
    - constructor
    - fromObject
    - fromStruct
    - toStruct
  - remove conditionIsValid method
  - in isValid, remove call to conditionIsValid

* In GroupHandlerTest.js,
  - added conditionStruct variable
  - in tests, removed "group.condition = condition" throughout
  - in setGroupCondition context
    - create group after condition in beforeEach
    - in "should update state" test
      - receive conditionStruct from getGroup
      - compare flattened condition struct to flattened expected condition

* In GroupTest.js,
  - removed expects checking conditionIsValid
  - removed condition from all Group constructors
  - removed test "Always present, condition must be a valid Condition instance"

* In OrchestrationHandlerTest.js,
  - in tests, removed "group.condition = condition" throughout
